### ISSUE TITLE
[resistance_defaults] and [terrain_defaults] are broken

### ISSUE THREAD

**Issue by CelticMinstrel** [wesnoth/wesnoth#5308]
Assuming the existence of a unit with a `type=sonic` attack, the following code works on 1.14.4 but does not work on 1.15 master:

```
[resistance_defaults]
	id=sonic
	default="(impact)"
[/resistance_defaults]
```

**What you should see:** If you place any unit, highlight it, and press D, it should show sonic resistance equal to their impact resistance.

**What you actually see:** Sonic resistance is not shown at all unless the unit has a movetype that explicitly specifies it.

Note: I didn't test, but it's possibly that `[terrain_defaults]` is similarly broken.

**Comment by Pentarctagon** [comment]
> 1.14.4

Did you mean 1.14.14?

**Comment by CelticMinstrel** [comment]
No, I haven't actually installed the latest stable version for several months, so it is indeed 1.14.4.

**Comment by Pentarctagon** [comment]
It looks like it broke somewhere between 1.14.7 and 1.14.9 (1.14.8 doesn't exist).

**Comment by Pentarctagon** [comment]
Removed the Blocker: Point label, seeing as this has been broken for multiple releases already without being noticed or reported.

**Comment by Pentarctagon** [comment]
@stevecotton it looks like this is caused by fcd0ceda37793f79ec59e76a2f621fb07af64c5b

**Comment by stevecotton** [comment]
Self-assigned and will investigate.

Does anyone have a good idea how to write a unit test for this, without adding an extra resistance to the units in all tests?

**Comment by stevecotton** [comment]
The commit that's just auto-linked above fixes part of this, which was caused by a string being "resistance" in one location and "resistances" in another. I'll continue debugging.

This now works (giving 70% defense):

        [resistance_defaults]
            id="special_res_for_test"
            default="30"
        [/resistance_defaults]
    
But using formulas still doesn't. It creates the new type but with the default value, not the value from arcane:

        [resistance_defaults]
            id="copy_of_arcane"
            default="(arcane)"
        [/resistance_defaults]

**Comment by stevecotton** [comment]
@Pentarctagon I'm wondering if it should be fixed or left as-is on the 1.14 branch. Either is going to be a change from a previous 1.14 release, meaning it could OOS a game - leaving it as-is means the OOS only occurs if a player has a version from before September 2019, which fixing it would give OOS if 1.14.15 was played with 1.14.14.

**Comment by stevecotton** [comment]
Okay, the same refactor was also the reason that formulas broke, again because I changed a line from `get_resistances()` to `", resistances"` - I ended up looking for "arcane", etc in the serialization of the movetype object, rather than in the child of the movetype object. For 1.15, fixing that could easily give an extra feature too:

@CelticMinstrel, would it be useful if movetype-patching formulas could access the full movetype? For example, resistance to "landslide" could be based on the movement cost for "mountain", or movement cost for "tangled metal junkyards with pointy bits" could depend on how easily the unit ignores "pierce" and "blade".

**Comment by soliton-** [comment]
Regressions should be fixed unless you have concrete issues that speak against it.

Access to the full movetype sounds like a good thing.

**Comment by Pentarctagon** [comment]
> Does anyone have a good idea how to write a unit test for this, without adding an extra resistance to the units in all tests?

You could write the test in a way that checks for a non-zero return value (thus running the test by itself) and then use an #ifdef that's only provided to that specific test to enable the [resistance_defaults]  Kinda hacky, but it should work I think.

> I'm wondering if it should be fixed or left as-is on the 1.14 branch. Either is going to be a change from a previous 1.14 release, meaning it could OOS a game - leaving it as-is means the OOS only occurs if a player has a version from before September 2019, while fixing it would give OOS if 1.14.15 was played with 1.14.14.

Given it's taken this long for someone to report it, I don't think there are many (if any) people using it on 1.14, so the risk of an OOS occurring is low.  Additionally it will be close to a year still until 1.16, so I'd prefer it be fixed in 1.14 rather than left not working on a stable branch until 1.16.

**Comment by stevecotton** [comment]
> You could write the test in a way that checks for a non-zero return value (thus running the test by itself) and then use an #ifdef that's only provided to that specific test to enable the [resistance_defaults] Kinda hacky, but it should work I think.

The mechanisms for setting #defs aren't supported by the test framework, at least, not without either doing a major rework of C++ code or having the test framework create a .cfg file. If there's a better way to hack it then please say, because this is the blocker for #5273.

> Given it's taken this long for someone to report it, I don't think there are many (if any) people using it on 1.14, so the risk of an OOS occurring is low. Additionally it will be close to a year still until 1.16, so I'd prefer it be fixed in 1.14 rather than left not working on a stable branch until 1.16.

Once it starts working again, I assume that at least one add-on will use it. @CelticMinstrel is this intended for an MP add-on? This is close to what #4751 is about, although it's a regression it's been non-functional so long that it's essentially a new API.

**Comment by Pentarctagon** [comment]
> The mechanisms for setting #defs aren't supported by the test framework, at least, not without either doing a major rework of C++ code or having the test framework create a .cfg file. If there's a better way to hack it then please say, because this is the blocker for #5273.

Are you talking about a Boost unit test or a WML unit test?

> Once it starts working again, I assume that at least one add-on will use it. @CelticMinstrel is this intended for an MP add-on? This is close to what #4751 is about, although it's a regression it's been non-functional so long that it's essentially a new API.

The bug report came from someone who was trying to add it, not from it being used in an existing add-on: https://forums.wesnoth.org/viewtopic.php?f=21&t=53619.

**Comment by stevecotton** [comment]
I was thinking of the WML unit tests, forgetting that the Boost ones can also do high-level tests (I forgot that they can see fully-built unit data).

**Comment by Pentarctagon** [comment]
That was campaign specific I thought, so would it work to have a test with something like:
#ifdef RES_DEFAULTS
[+units]
    [resistance_defaults]
        id=electric
        default=100
        undeadfly=130
        undeadfoot=130
        undeadspirit=90
    [/resistance_defaults]
[/units]
#endif

and then provide RES_DEFAULTS as a define for that test?  Since from the issue it sounded the issue was that the tests exited before getting to anything campaign-specific, rather than they didn't support defines at all.

**Comment by stevecotton** [comment]
Added a lot more detail to #5273, the difficulties implementing it are more about preprocessing than about campaigns.

For fixing in this issue: the schema only allows one [resistance_defaults], while it's clearly a tag that should allow multiple instances.

For the Boost-based unit tests, the resistance data didn't seem to be available. I'd rather use a WML-based test, so not exploring further with this:

```
/**
 * Check that build_unit_type(, unit_type::FULL) includes the resistances.
 */
BOOST_AUTO_TEST_CASE( additional_resistance_type ) {
	config game_config(test_utils::get_test_config());

	config elf_config;
	elf_config["id"]="Elvish Archer";
	elf_config["random_traits"] = false;
	elf_config["animate"]=false;
	unit_type elf_type(elf_config);

	unit_types.build_unit_type(elf_type, unit_type::FULL);

	// Elvish Archers have a -10% malus against arcane. If the first one of
	// these fails then the basic unit data isn't available, and the failure is
	// before we get to testing [resistance_defaults]'s effects.
	BOOST_CHECK_EQUAL(elf_type.resistance_against("arcane", true), 90);
	// \todo: "arcane" is currently failing. This was intended to be expanded to
	// test [resistance_defaults], once the basic "arcane" value worked.
}
```

**Comment by CelticMinstrel** [comment]
> would it be useful if movetype-patching formulas could access the full movetype?

I didn't think of that when I implemented it, but I don't really see any reason why not?

> is this intended for an MP add-on?

I suppose it should work in MP, sure. I'd say it's _intended_ more for campaigns though.

**Comment by CelticMinstrel** [comment]
So this is to be backported now, is that correct?

**Comment by stevecotton** [comment]
It's one I've been procrastinating about responding to. I'm not sure whether to

* backport and change to the exact syntax that 1.14.4 supported
* backport including the extra syntax of 1.15.x
* argue further that, even with the 1.14.4 syntax, it would effectively be a new API if re-added to 1.14.16

**Comment by CelticMinstrel** [comment]
I think the third option is probably wrong, at least. It _used_ to work in 1.14 so shouldn't it continue to work?

**Comment by Pentarctagon** [comment]
Fixing something that used to work in a previous version of the same stable release series would never count as adding a new API.  I don't think it's relevant whether a bug like this is noticed soon after it's introduced or if it's noticed much later (as is the case here) - it's a breakage that occurred between one release of a stable series and another and should be fixed.

As far as whether to use the new syntax or not, I think that depends on whether the new syntax would break any existing code in 1.14.  APIs aren't supposed to have their syntax changed to be incompatible with previous 1.14 releases - one of the main reasons for having a stable branch at all is *not* doing that sort of thing.

So I think the correct option is to backport the fix using the existing syntax in 1.14.  If any brand new tags/attributes have been added as part of fixing the issue, then that's fine too.

**Comment by stevecotton** [comment]
Thanks for the feedback; in that case it's simple to pull the code from 1.15 back to 1.14, after first making 1.15 recognise `[movement]` as an alias for `[movement_costs]`.

* Formulas already supported by 1.14.8 will still work, for example `(shallow_water)`
* Formulas that access the full movetype will expect the 1.15 name, for example `(movement_costs.shallow_water)`

**Comment by stevecotton** [comment]
Fixed with the merge of #5583.
